--- ff.c	Tue Jul 22 00:42:24 2025
+++ ff16p1.c	Sat Sep 13 18:37:13 2025
@@ -1,4 +1,4 @@
 /*----------------------------------------------------------------------------/
-/  FatFs - Generic FAT Filesystem Module  R0.16                               /
+/  FatFs - Generic FAT Filesystem Module  R0.16 w/patch 1                     /
 /-----------------------------------------------------------------------------/
 /
@@ -42,4 +42,6 @@
 #define MAX_FAT32	0x0FFFFFF5		/* Max FAT32 clusters (not defined in specs, practical limit) */
 #define MAX_EXFAT	0x7FFFFFFD		/* Max exFAT clusters (differs from specs, implementation limit) */
+#define MIN_FAT12	32				/* Min FAT12 clusters (Not defined in specs, implementation limit) */
+#define MIN_VOLUME	64				/* Min volume sectors (Not defined in specs, implementation limit) */
 
 
@@ -3393,5 +3395,5 @@
 			&& (UINT)fs->win[BPB_NumFATs] - 1 <= 1		/* Properness of number of FATs (1 or 2) */
 			&& ld_16(fs->win + BPB_RootEntCnt) != 0		/* Properness of root dir size (MNBZ) */
-			&& (ld_16(fs->win + BPB_TotSec16) >= 128 || ld_32(fs->win + BPB_TotSec32) >= 0x10000)	/* Properness of volume size (>=128) */
+			&& (ld_16(fs->win + BPB_TotSec16) >= MIN_VOLUME || ld_32(fs->win + BPB_TotSec32) >= 0x10000)	/* Properness of volume size */
 			&& ld_16(fs->win + BPB_FATSz16) != 0) {		/* Properness of FAT size (MNBZ) */
 				return 0;	/* It can be presumed an FAT VBR */
@@ -3618,4 +3620,5 @@
 		if (nclst <= MAX_FAT16) fmt = FS_FAT16;
 		if (nclst <= MAX_FAT12) fmt = FS_FAT12;
+		if (nclst <= MIN_FAT12) fmt = 0;
 		if (fmt == 0) return FR_NO_FILESYSTEM;
 
@@ -6158,5 +6161,5 @@
 		}
 	}
-	if (sz_vol < 128) LEAVE_MKFS(FR_MKFS_ABORTED);	/* Check if volume size is >=128 sectors */
+	if (sz_vol < MIN_VOLUME) LEAVE_MKFS(FR_MKFS_ABORTED);	/* Check if volume size is not too small */
 
 	/* Now start to create an FAT volume at b_vol and sz_vol */
@@ -6390,5 +6393,6 @@
 
 			/* Determine number of clusters and final check of validity of the FAT sub-type */
-			if (sz_vol < b_data + pau * 16 - b_vol) LEAVE_MKFS(FR_MKFS_ABORTED);	/* Too small volume? */
+			if (sz_vol < b_data + pau * MIN_FAT12 - b_vol) LEAVE_MKFS(FR_MKFS_ABORTED);	/* Too small volume for this configuration? */
+
 			n_clst = ((DWORD)sz_vol - sz_rsv - sz_fat * n_fat - sz_dir) / pau;
 			if (fsty == FS_FAT32) {
